rust/driver-ipc/src/client.rs

```diff
@@ -1,180 +1,157 @@
-use std::{
-    convert::Infallible,
-    io,
-    sync::{
-        atomic::{AtomicBool, Ordering},
-        Arc,
-    },
-};
+use std::{io, sync::Arc, time::Duration};
 
 use log::error;
 use serde::Serialize;
 use tokio::{
-    net::windows::named_pipe::{ClientOptions, NamedPipeClient, PipeMode},
-    sync::{
-        mpsc::{error::TryRecvError, unbounded_channel, UnboundedReceiver, UnboundedSender},
-        Mutex,
-    },
+    net::windows::named_pipe,
+    sync::{broadcast, Notify},
     task,
+    time::timeout,
 };
-use winreg::{
-    enums::{HKEY_CURRENT_USER, KEY_WRITE},
-    RegKey,
-};
+use tokio_stream::{Stream, StreamExt};
 
-use crate::{
-    ClientCommand, DriverCommand, EventCommand, Id, IpcError, Monitor, ReplyCommand,
-    RequestCommand, Result,
-};
+use crate::*;
 
 // EOF byte used to separate messages
-const EOF: u8 = 0x4;
-
-/// A thin api client over the driver api with all the essential api.
-/// Does the bare minimum required. Does not track state
-#[derive(Debug, Clone)]
-pub struct Client(Arc<Inner>);
-
+pub(crate) const EOF: u8 = 0x4;
+
+/// Client for interacting with the Virtual Display Driver.
+///
+/// Connects via a named pipe to the driver.
+///
+/// You can send changes to the driver and receive continuous events from it.
+///
+/// It is save to clone this client. The connection is shared between all
+/// copies.
 #[derive(Debug)]
-struct Inner {
-    client: NamedPipeClient,
-    user_is_event: AtomicBool,
-    event_recv: Mutex<UnboundedReceiver<EventCommand>>,
-    client_recv: Mutex<UnboundedReceiver<ReplyCommand>>,
+pub struct Client {
+    client: Arc<named_pipe::NamedPipeClient>,
+    command_rx: broadcast::Receiver<ClientCommand>,
+    abort_receiver: Arc<Notify>,
 }
 
 impl Client {
-    /// connect to pipe virtualdisplaydriver
-    pub async fn connect() -> Result<Self> {
-        Self::connect_to("virtualdisplaydriver").await
+    /// Connect to driver on pipe with default name.
+    ///
+    /// The default name is [DEFAULT_PIPE_NAME].
+    pub fn connect() -> Result<Self> {
+        Self::connect_to(DEFAULT_PIPE_NAME)
     }
 
-    // choose which pipe name you connect to
-    // pipe name is ONLY the name, only the {name} portion of \\.\pipe\{name}
-    pub async fn connect_to(name: &str) -> Result<Self> {
-        let client = ClientOptions::new()
+    /// Connect to driver on pipe with specified name.
+    ///
+    /// `name` is ONLY the {name} portion of \\.\pipe\{name}.
+    pub fn connect_to(name: &str) -> Result<Self> {
```

Comment:
The two client methods are a little misleading, cause even though you can create a named pipe using sync code, it actually requires a reactor to be running. In pure sync code it would result in
`there is no reactor running, must be called from the context of a Tokio 1.x runtime`

We should have both of these be `async` to reduce the possibility anyone would run into the error (though it might be a slightly remote chance since if they're using this they're likely already running tokio in the first place, but regardless)